text/0000-reloadable-process-types.md

```diff
@@ -0,0 +1,36 @@
+# Reloadable Process Types
+
+## Summary
+
+Provide a utility buildpack to support language buildpacks in creating reloadable process types for their run containers.
+
+## Motivation
+
+Usage of dev orchestration tools like Tilt and Skaffold are on the rise. These tools aim to make  developing distributed software that runs on Kubernetes easier.  Offering the capability to develop inside the container via live reload features.  With this, changes are streamed from the developer’s workstation into the running container whereupon the init process is restarted.  The dev orchestrator, of course, handles most of this by building the changes and copying them in the container but ultimately the running container needs the ability to restart its entrypoint process for the changes to take effect.
+
+Today buildpacks are unable to participate in these live reload development workflows but, moving forward, we would very much like them to be able to.
+
+## Detailed Explanation
+
+We propose creating a utility buildpack that packages the watchexec command line utility for restarting arbitrary processes.
```

Comment:
So, the proposal is simply to create a `watchexec` buildpack, similar to other dependency-providing buildpacks like [Paketo Yarn](https://github.com/paketo-buildpacks/yarn) or [Paketo Bundler](https://github.com/paketo-buildpacks/bundler)?

Reply:
I'm not sure it's that simple. The existing utility buildpacks are basically self-sufficient. They get invoke, do something and the other buildpacks don't care. Think Procfile or env-variables. They are setting process types or env variables on the image, other buildpacks don't need to be aware this is happening.

I think this proposal is a little more involved because the language buildpack would have to opt-in and modify its process types if watchexec is enabled, since there's no way for `watchexec` to emit the necessary process types (at least I don't think it's possible).

Reply:
🤔  I see. Seems like language buildpacks would be able to "opt in" to adding the reload feature once the watchexec buildpack exists.

I'm wondering if, for the sake of simplicity, this RFC should just propose and outline the functionality of the watchexec buildpack. Or is part of the effort here to align the Paketo buildpacks to _all_ have reloadable processes support that works roughly the same? That's probably a question for @paulcwarren.

Reply:
There has been some discussion in the past about whether it should be possible for buildpacks to see what process types have been defined by other buildpacks (technically, it's possible to read other buildpacks' launch.toml but not defined by the spec). I wonder if having such a mechanism would make this proposal easier.

Reply:
What @dmikusa-pivotal says basically.  And also see Emily's [comment above](https://github.com/paketo-buildpacks/rfcs/pull/71#discussion_r647666558)

Reply:
@paulcwarren can you add a bit to clarify this? I see three points that need to be described.

1. Clarify that the watchexec buildpack is just providing the binary, nothing more. That makes it very simple.
2. Clarify what another buildpack would need to do to opt-in. Something to the effect of watch for a variable like Emily mentions in her comment and then modify the start command to use watchexec.
3. I think you'd like to start with Java support for this, so indicate that as well in the proposal.

Reply:
Done